File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# 1.순서를 바꾸지 말고 더하거나 빼서 타겟 넘버를 만든다
# 2. visit은 필요 없으며 numbers를 따라 +1 -1 을 하면서 target과 맞는지 확인해야 함
from collections import deque
answer=0
def DFS(numbers,target,num,current):
    global answer
    if num == len(numbers) and current==target:
        answer+=1
        return
    elif num == len(numbers):
        return
    DFS(numbers,target,num+1,current+numbers[num])
    DFS(numbers,target,num+1,current-numbers[num])
def solution(numbers, target):
    DFS(numbers,target,0,0)
    
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
      static int answer =0;
    public int solution(int[] numbers, int target) {

        dfs(numbers, 0, target, 0);

        return answer;

    }
    private void dfs(int[] numbers, int result, int target, int i) {
        if (i == numbers.length) {
            if (result == target) {
                answer++;
            }
            return;
        }
        dfs(numbers, result - numbers[i], target, i + 1);
        dfs(numbers, result + numbers[i], target, i +1);
    }
}